src/metabase/models/collection.clj

```diff
@@ -535,39 +535,34 @@
       :id [:in ancestor-ids]
       {:order-by [:location]})))
 
-(mi/define-simple-hydration-method ^:private ancestors
+(mi/define-simple-hydration-method ancestors
   :ancestors
   "Fetch ancestors (parent, grandparent, etc.) of a `collection`. These are returned in order starting with the
   highest-level (e.g. most distant) ancestor."
   [collection]
   (ancestors* collection))
 
-(mu/defn- effective-ancestors* :- [:sequential [:or RootCollection (ms/InstanceOf Collection)]]
-  [collection :- CollectionWithLocationAndIDOrRoot]
-  (if (collection.root/is-root-collection? collection)
+(mu/defn- effective-ancestors*
+  "Given a collection, return the effective ancestors of that collection."
+  [collection :- [:maybe CollectionWithLocationOrRoot]]
+  (if (or (nil? collection)
+          (collection.root/is-root-collection? collection))
     []
-    (filter mi/can-read? (cons (root-collection-with-ui-details (:namespace collection)) (ancestors collection)))))
-
-(mi/define-simple-hydration-method effective-ancestors
+    (some->> (effective-location-path collection)
+             location-path->ids
+             (map (collection-id->collection))
+             (map #(select-keys % [:name :id :personal_owner_id :type]))
+             (map #(t2/instance :model/Collection %))
+             (cons (root-collection-with-ui-details (:namespace collection)))
+             (filter mi/can-read?))))
+
+(mi/define-batched-hydration-method effective-ancestors
   :effective_ancestors
-  "Fetch the ancestors of a `collection`, filtering out any ones the current User isn't allowed to see. This is used
-  in the UI to power the 'breadcrumb' path to the location of a given Collection. For example, suppose we have four
-  Collections, nested like:
-
-    A > B > C > D
-
-  The ancestors of D are:
-
-    [Root] > A > B > C
-
-  If the current User is allowed to see A and C, but not B, `effective-ancestors` of D will be:
-
-    [Root] > A > C
-
-  Thus the existence of C will be kept hidden from the current User, and for all intents and purposes the current User
-  can effectively treat A as the parent of C."
-  [collection]
-  (effective-ancestors* collection))
+  "Efficiently hydrate the `:effective_ancestors` of collections."
```

Comment:
Does `effective-ancestors*` do any DB calls or is it just using cached data to compute the ancestors? I was initially wondering if this could be an N+1 so just want to confirm. If it's not maybe that could be a little clearer in the docstring/comments.